crates/sui-core/src/authority/consensus_quarantine.rs

```diff
@@ -330,3 +334,647 @@ impl ConsensusCommitOutput {
+            .get_all_user_signatures_for_checkpoints()
+            .expect("load user signatures for checkpoints cannot fail");
+
+        Self {
+            shared_version_assignments: shared_version_assignments.into_iter().collect(),
+            deferred_transactions: Mutex::new(deferred_transactions),
+            user_signatures_for_checkpoints: Mutex::new(user_signatures_for_checkpoints),
+        }
+    }
+
+    // Used to read pre-existing shared object versions from the database after a crash.
+    // TODO: remove this once all nodes have upgraded to data-quarantining.
+    fn get_all_shared_version_assignments(
+        epoch_start_configuration: &EpochStartConfiguration,
+        tables: &AuthorityEpochTables,
+    ) -> Vec<(
+        TransactionKey,
+        Vec<(ConsensusObjectSequenceKey, SequenceNumber)>,
+    )> {
+        if epoch_start_configuration.use_version_assignment_tables_v3() {
+            tables
+                .assigned_shared_object_versions_v3
+                .safe_iter()
+                .collect::<Result<_, _>>()
+                .expect("db error")
+        } else {
+            tables
+                .assigned_shared_object_versions_v2
+                .safe_iter()
+                .collect::<Result<Vec<_>, _>>()
+                .expect("db error")
+                .into_iter()
+                .map(|(key, value)| {
+                    (
+                        key,
+                        value
+                            .into_iter()
+                            .map(|(id, v)| ((id, SequenceNumber::UNKNOWN), v))
+                            .collect::<Vec<_>>(),
+                    )
+                })
+                .collect()
+        }
+    }
+}
+
+/// ConsensusOutputQuarantine holds outputs of consensus processing in memory until the checkpoints
+/// for the commit have been certified.
+pub(crate) struct ConsensusOutputQuarantine {
+    // Output from consensus handler
+    output_queue: VecDeque<ConsensusCommitOutput>,
+
+    // Highest known certified checkpoint sequence number
+    highest_executed_checkpoint: CheckpointSequenceNumber,
+
+    // Checkpoint Builder output
+    builder_checkpoint_summary:
+        BTreeMap<CheckpointSequenceNumber, (BuilderCheckpointSummary, CheckpointContents)>,
+
+    builder_digest_to_checkpoint: HashMap<TransactionDigest, CheckpointSequenceNumber>,
+
+    // Any un-committed next versions are stored here.
+    shared_object_next_versions: RefCountedHashMap<ConsensusObjectSequenceKey, SequenceNumber>,
+
+    // The most recent congestion control debts for objects. Uses a ref-count to track
+    // which objects still exist in some element of output_queue.
+    congestion_control_randomness_object_debts:
+        RefCountedHashMap<ObjectID, CongestionPerObjectDebt>,
+    congestion_control_object_debts: RefCountedHashMap<ObjectID, CongestionPerObjectDebt>,
+
+    processed_consensus_messages: RefCountedHashMap<SequencedConsensusTransactionKey, ()>,
```

Comment:
I think we talked in the meeting too, but it seems that processed_consensus_messages does not need to be `RefCountedHashMap`? When first commit encountered message is written to disk, the message can be removed from memory, since it is on the disk anyway. Moreover, I think only one commit can ever put SequencedConsensusTransactionKey in this map?

Reply:
I tried to fix this, but it turns out that for deferred transactions, we insert them twice. Once when they arrive for the first time, and once again when we load them from the deferral queue. It would be nice to fix this, but its quite a chore to do.